### PR TITLE
Fal 13

### PR DESCRIPTION
**Pull request by willpennell**
Login API working, unit tested, custom Repository code test and integration tested with DB, end-to-end with manual Postman  testing.

### CODE REVIEW

File: login/src/test/java/com/teamfalcon/login/service/LoginServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+        when(userRepository.findByUsername(any(String.class))).thenReturn(Optional.of(validUser));
+        when(userTokenRepository.findByUserId(any(Integer.class))).thenReturn(Optional.of(returnedUserToken));
+        when(userTokenRepository.save(any(TokenEntity.class))).thenReturn(any(TokenEntity.class));
+
+        LoginResponseDTO actualResponse = loginService.authoriseLogin(loginRequestBodyDTO);
+        Boolean expectedSuccess = true;
+
+        assertNotEquals(returnedUserToken.getToken(), actualResponse.getToken());
+        assertEquals(expectedSuccess, actualResponse.getSuccess());
```
Comment (VincentFleetwood) [inline login/src/test/java/com/teamfalcon/login/service/LoginServiceTest.java:159]: assertTrue

File: login/src/test/java/com/teamfalcon/login/service/LoginServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+        when(userRepository.findByUsername(any(String.class))).thenReturn(Optional.of(validUser));
+        when(userTokenRepository.findByUserId(any(Integer.class))).thenReturn(Optional.of(returnedUserToken));
+
+
+        LoginResponseDTO actualResponse = loginService.authoriseLogin(loginRequestBodyDTO);
+        Boolean expectedSuccess = true;
+
+        assertNotNull(actualResponse.getToken());
+        assertEquals(returnedUserToken.getToken(), actualResponse.getToken());
```
Comment (VincentFleetwood) [inline login/src/test/java/com/teamfalcon/login/service/LoginServiceTest.java:185]: try and assert a value not the result of a getter please

File: login/src/test/java/com/teamfalcon/login/testmodels/TestUserEntityFactory.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
+package com.teamfalcon.login.testmodels;
+
+import com.teamfalcon.login.model.LoginRequestBodyDTO;
+import com.teamfalcon.login.model.UserEntity;
+
+public class TestUserEntityFactory {
```
Comment (VincentFleetwood) [inline login/src/test/java/com/teamfalcon/login/testmodels/TestUserEntityFactory.java:6]: these classes are often called Fixtures in test code

**Review by VincentFleetwood — CHANGES_REQUESTED** [review]

File: login/src/main/java/com/teamfalcon/login/exceptions/ExceptionMessages.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.teamfalcon.login.exceptions;
```
[the remark this answers is not in the document]
Comment (willpennell) [reply login/src/main/java/com/teamfalcon/login/exceptions/ExceptionMessages.java:1]: perfect, will do!

File: login/src/main/java/com/teamfalcon/login/service/LoginServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        if (tokenEntity.getExpiryDate().isBefore(LocalDateTime.now())) {
+            tokenEntity = updateTokenEntity(tokenEntity);
+        }
+
+        return getLoginResponseDTO(tokenEntity);
+    }
+
+    private void validateLoginRequestUsername(LoginRequestBodyDTO loginRequestBodyDTO) {
+        if (loginRequestBodyDTO.getUsername().isEmpty() || loginRequestBodyDTO.getUsername().isBlank()) {
+            throw new EntityNotFoundException(ENTITY_NOT_FOUND_ERROR_MESSAGE);
```
[the remark this answers is not in the document]
Comment (willpennell) [reply login/src/main/java/com/teamfalcon/login/service/LoginServiceImpl.java:54]: OK great!

File: login/src/main/java/com/teamfalcon/login/model/UserEntity.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+@NoArgsConstructor
+public class UserEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Integer id;
+    private String username;
+    private String passwordHash;
+    private Integer failedLoginAttempts;
+    private Boolean isDeleted;
```
[the remark this answers is not in the document]
Comment (willpennell) [reply login/src/main/java/com/teamfalcon/login/model/UserEntity.java:27]: sure!

**Comment by Jomm0** [comment]
All looks good to me; can't see anything additional which hasn't already been covered by Vince, which have now been sorted.

**Review by Jomm0 — APPROVED** [review]
All changes requested have been implemented.

**Review by VincentFleetwood — APPROVED** [review]

**Pull request merged**